### ISSUE TITLE
Importing epf preferences file from STS3 doesn't work

### ISSUE THREAD

**Issue by thejeff77** [spring-projects/sts4#134]
It appears that exporting preferences from STS3 and importing them again in STS4 doesn't work. 

I've tried import/export of preferences.
I've also tried every suggestion in this post: https://stackoverflow.com/questions/2078476/how-to-share-eclipse-configuration-over-different-workspaces
I've also tried just copying over the entire STS3 workspace to the STS4 workspace. No luck. 

I know the preferences haven't transferred because none of the workbench settings, compiler options, theme options, validation options, aspect options, or formatting options copied over. I'm still opening the same project with all of the same .settings and .metadata folders in it.. so its not a question of project level settings either. 

Please fix this! It will make migration to a newer version of STS much less painful.

**Comment by thejeff77** [comment]
Well.. it appears that updating STS blew away the settings for both versions. As usual: user error.

It would be pretty cool if STS4 migrated the settings instead of blowing everything away for a new format.

**Comment by martinlippert** [comment]
I agree, destroying preferences is a pretty bad behavior when it comes to migrating to a new version, absolutely agree. I wonder why this happens though and which preferences are affected - also to identify what causes this behavior. Need to investigate.

Just to make sure I understand things correctly: you opened an old workspace that you previously used with STS3 with a new and fresh install of STS4 and suddenly your preferences where gone? Or is this exclusively about exporting/importing preferences?

**Comment by thejeff77** [comment]
I am not sure if the import/export works - that was a red herring as I had not yet realized my project settings were reset, so I had been exporting the reset settings and importing them again. 

When I installed the new STS and opened the project, there was a message about upgrading the project or something - which makes sense. The same thing happens with other IDEs like Visual Studio - it has to migrate to the new project schema. It appears to have reset all of my preferences after I hit confirm in the dialog.

**Comment by thejeff77** [comment]
I wish I did, otherwise the import may have worked. Also its a company project so I'm not sure if I could provide it anyway. 

Luckily, this should be pretty straightforward to reproduce, assuming the issue is that STS does not migrate settings on an upgrade.

1. Use STS3 to import any maven project. 
2. In settings, configure the dark theme, validation settings (ignore html files or something), warnings, **basically anything that isn't default**. You might change a bunch of stuff and write down what was changed.
3. Install STS4 and open up the same project in a new workspace. If it is working you'll receive a prompt about a project migration of some sort.
4. Check settings - all reset back to default.

Basically the steps you'd walk through to test a project migration in an IDE if you're a developer working on the feature.

In case it matters, I'm using a mac.

**Comment by martinlippert** [comment]
Okay, here is what I tried and my observations so far:

- created a fresh workspace using STS 3.9.6 (based on Eclipse 4.9)
- created a new project (from the guides using the Rest Service guide)
- changed various settings (global: dark theme + no automatic updates, project-specific: various compiler settings)
- ran a fresh STS 4.0.1 (based on Eclipse 4.9)
- created a fresh workspace
- imported the project using "Import Existing Projects into Workspace" -> no dialog showed up
- the global settings were not applied (as expected, since those are workspace-related settings, not project-related)
- the project-specific settings were still there

Tried to re-open the same workspace -> everything just worked, including the global setting changes

Then I created a fresh workspace using an STS 3.9.5 (on Eclipse 4.8) and created a similar project again in that workspace, changed the various settings again

-> when importing this project using "Import Existing Projects into Workspace", the project got imported correctly, again no global setting changes applied, but all the project-specific ones made it

-> when opening the old workspace (instead of using a new one and importing the project), I got a dialog telling me that this workspace is from an older Eclipse version and will be migrated to the newer version
-> pressed "OK"
-> global settings got applied just fine
-> project-local settings made it, too

The only exception to the global settings are the changes that I made to the "Update automatically", which might be related to the fact that we have custom settings for that defined in the STS product definition.

So far this "keeping the preferences" seems to work as expected. Workspace-related settings like the dark theme make it when you re-open the same workspace. Project-specific settings always made it. So I am somewhat unsuccessful trying to reproduce this problem.

Do you have any idea how I might be able to reproduce this? Any more concrete input (like concrete STS versions that you used, which settings got lost, etc.) might be helpful here.

**Comment by martinlippert** [comment]
There is one thing that I am aware of:

Some Java content-assist related settings might get lost when re-opening the same workspace with STS4. This is caused by the content-assist preferences being manipulated automatically by the code recommenders extension that was installed in STS3 by default. It disabled the regular Java content-assist computers and instead enabled its own ones. In STS4 we don't install the code recommenders extension by default. So the code-recommenders-specific content-assist computers are not around anymore, but the regular Java ones are not automatically re-enabled again. You have to re-enable them or reset to the defaults to get the Java content-assist features back.

**Comment by thejeff77** [comment]
@martinlippert thank you for the information and being so thorough. From what you've said, it is possible I've made a mistake or have done something off of the beaten path as far as flow, or have a special case of some sort. 

I'm switching back to STS3 for the time being, and will do this migration again in the next week or two, this time being careful to document what happens and observe if I encounter any special cases. As I've already done this migration before, I know what to pay attention to more than last time.

**Comment by martinlippert** [comment]
I don't think your lost of preferences was a user error, to be honest. I can totally imagine that there is a migration path from STS3 to STS4 that triggers a certain bug or a missing piece that causes this. The challenge here is to identify that path and the bug that gets triggered by that path. So please keep a close eye and record the steps carefully so that we can maybe identify this issue, reproduce it, and, of course, fix it... :-)

I would close this issue for now, but please add your comments once you got the chance to do the migration again and we re-open this issue again - to continue the investigation.
